### PR TITLE
Revise PacketBufferWriter interface

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

The `PacketBufferWriter` constructor that takes a size and allocates the
`PacketBuffer` internally is awkward for some cases, and useless for the
case of an existing `PacketBufferHandle`.

#### Summary of Changes

- `PacketBufferWriter()` takes ownership of an existing `PacketBufferHandle`
  (same as `PacketBufferTLVWriter`).
- Added a `Finalize()` overload that matches that of `PacketBufferTLVWriter`.

### CODE REVIEW

File: src/system/SystemPacketBuffer.h [hunk 1]
```diff
@@ -730,10 +730,9 @@ class PacketBufferWriterBase : public Writer
      *  @param[in]  aAvailableSize  Length bound of the BufferWriter.
      *  @param[in]  aReservedSize   Reserved packet buffer space for protocol headers; see \c PacketBufferHandle::New().
      */
-    PacketBufferWriterBase(size_t aAvailableSize, uint16_t aReservedSize = CHIP_SYSTEM_CONFIG_HEADER_RESERVE_SIZE) :
-        Writer(nullptr, 0)
+    PacketBufferWriterBase(System::PacketBufferHandle && aPacket) : Writer(aPacket->Start(), aPacket->AvailableDataLength())
```
Comment (kpschoedel) [inline src/system/SystemPacketBuffer.h:733]: Looks like this still needs a length (possibly optional; checking the cases), for the case that the pool-backed buffer is actually larger than requested.

**Review by mspang — APPROVED** [review]

File: src/transport/NetworkProvisioning.cpp [hunk 2]
```diff
@@ -264,7 +264,7 @@ CHIP_ERROR NetworkProvisioning::SendThreadCredentials(const DeviceLayer::Interna
         4;                  // threadData.ThreadChannel, threadData.FieldPresent.ThreadExtendedPANId,
                             // threadData.FieldPresent.ThreadMeshPrefix, threadData.FieldPresent.ThreadPSKc
     /* clang-format on */
-    Encoding::LittleEndian::PacketBufferWriter bbuf(credentialSize);
+    Encoding::LittleEndian::PacketBufferWriter bbuf(System::PacketBufferHandle::New(credentialSize));
```
Comment (bzbarsky-apple) [inline src/transport/NetworkProvisioning.cpp:267]: This should be a `MessagePacketBuffer` too.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
I'm OK with a followup for the MessagePacketBuffer bits, so that we don't combine behavior changes with refactoring in a single PR.  Up to you.

But we do need to sort out the way we get the pointer+length out of the buffer.

File: src/system/SystemPacketBuffer.h [hunk 3]
```diff
@@ -757,6 +756,22 @@ class PacketBufferWriterBase : public Writer
      */
     System::PacketBufferHandle Finalize() { return PacketBufferWriterUtil::Finalize(*this, mPacket); }
 
+    /**
+     * Finish the writing to the buffer and release ownership of the underlying PacketBuffer.
+     *
+     * This signature matches `PacketBufferTLVWriter::Finalize()`.
```
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/system/SystemPacketBuffer.h:762]: The original signature is more useful in practice; almost all Writer users immediately pass the buffer somewhere else.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/system/SystemPacketBuffer.h:762]: OK, but why add the new one then?

File: src/protocols/bdx/BdxTransferSession.cpp [hunk 4]
```diff
@@ -25,19 +25,13 @@ constexpr size_t kStatusReportMinSize = 2 + 4 + 2; ///< 16 bits for GeneralCode,
 CHIP_ERROR WriteToPacketBuffer(const ::chip::bdx::BdxMessage & msgStruct, ::chip::System::PacketBufferHandle & msgBuf)
 {
     size_t msgDataSize = msgStruct.MessageSize();
-    ::chip::Encoding::LittleEndian::PacketBufferWriter bbuf(msgDataSize);
+    ::chip::Encoding::LittleEndian::PacketBufferWriter bbuf(::chip::System::PacketBufferHandle::New(msgDataSize));
```
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/protocols/bdx/BdxTransferSession.cpp:28]: I'll add them here so they don't get overlooked.

File: src/system/SystemPacketBuffer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/system/SystemPacketBuffer.h:733]: Correct, I started off intended the former, but I think the latter is more useful.

File: src/system/SystemPacketBuffer.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (kpschoedel) [reply src/system/SystemPacketBuffer.h:762]: Consistency with `PacketBufferTLVWriter` seemed like a good idea at the time.

File: src/protocols/bdx/BdxTransferSession.cpp [hunk 5]
```diff
@@ -25,7 +26,7 @@ constexpr size_t kStatusReportMinSize = 2 + 4 + 2; ///< 16 bits for GeneralCode,
 CHIP_ERROR WriteToPacketBuffer(const ::chip::bdx::BdxMessage & msgStruct, ::chip::System::PacketBufferHandle & msgBuf)
 {
     size_t msgDataSize = msgStruct.MessageSize();
-    ::chip::Encoding::LittleEndian::PacketBufferWriter bbuf(msgDataSize);
+    ::chip::Encoding::LittleEndian::PacketBufferWriter bbuf(chip::MessagePacketBuffer::New(msgDataSize));
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.cpp:29]: Should this also pass `msgDataSize` as second arg?  Or document why we don't care if this overallocates?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**